### PR TITLE
Handle fully compressed messages

### PR DESCRIPTION
**Pull request by bigglesss**
- [x] The part of the SMSG_COMPRESSED_UPDATE_OBJECT test that compares the original raw packet to the newly written + compressed packet fails due to a size comparison. If I decompress both packets, they result in the exact same data, but the original vmangos packet is considerably smaller (by about 20 bytes) than the flate2 compressed data, despite having a header indicating that ZLIB level 1 was used.

### CODE REVIEW

File: wow_message_parser/src/rust_printer/mod.rs [hunk 1]
```diff
@@ -160,10 +160,11 @@ impl Writer {
     pub(crate) fn variable_size(
         &mut self,
         name: impl AsRef<str>,
+        function_name: &str,
```
Comment (gtker) [inline wow_message_parser/src/rust_printer/mod.rs:163]: Why is this necessary?

Comment (gtker) [reply wow_message_parser/src/rust_printer/mod.rs:163]: Ah, OK, makes sense. I completely missed that.

File: wow_message_parser/src/doc_printer/container.rs [hunk 2]
```diff
@@ -434,8 +434,26 @@ fn print_container_examples(s: &mut DocWriter, e: &Container, o: &Objects) {
 
         let mut values = HashMap::new();
 
-        for m in e.members() {
-            print_container_example_member(s, e, m, &mut bytes, &mut values, o, e.tags(), "");
+        if e.tags().compressed() {
```
[the remark this answers is not in the document]
Comment (bigglesss) [reply wow_message_parser/src/doc_printer/container.rs:437]: `bytes` is an iterator, so I can't return it from an if case (as the original Vec: decoded_bytes is dropped)

File: wow_message_parser/src/rust_printer/structs/print_common_impls/print_read.rs [hunk 3]
```diff
@@ -835,6 +835,18 @@ pub(crate) fn print_read(s: &mut Writer, e: &Container, o: &Objects, prefix: &st
         return;
     }
 
+    // For fully compressed messages, replace the writer with a ZLibDecoder.
+    if e.tags().compressed() {
+        // Fully compressed messages always start with a u32 containing the decompressed size.
+        // We don't care about that, so we just ignore it.
```
Comment (gtker) [inline wow_message_parser/src/rust_printer/structs/print_common_impls/print_read.rs:841]: Can't `flate` pre allocate based on it?

File: wowm_language/src/docs/smsg_compressed_update_object.md [hunk 4]
```diff
@@ -23,5 +25,120 @@ SMSG have a header of 4 bytes.
...
 ### Body
 
-The body for this message has not been implemented yet.
+| Offset | Size / Endianness | Type | Name | Description | Comment |
+| ------ | ----------------- | ---- | ---- | ----------- | ------- |
+| 0x04 | 4 / Little | u32 | amount_of_objects |  |  |
+| 0x08 | 1 / - | u8 | has_transport |  |  |
+| 0x09 | ? / - | [Object](object.md)[amount_of_objects] | objects |  |  |
 
+### Examples
```
Comment (gtker) [inline wowm_language/src/docs/smsg_compressed_update_object.md:34]: I'm unsure of how to deal with the examples here. On one hand it's useful to see the uncompressed data explained but on the other it gives an incorrect view of what's actually going on.

I also noticed that the example doesn't have the uncompressed size `u32`. Maybe generate example text for both the compressed and uncompressed data?

**Review by gtker — CHANGES_REQUESTED** [review]
Looking good, does the message work if you use it instead of `SMSG_UPDATE_OBJECT` in the example, or how do you know that it's correct?

**Comment by bigglesss** [comment]
> Looking good, does the message work if you use it instead of `SMSG_UPDATE_OBJECT` in the example, or how do you know that it's correct?

I hadn't tried that, but I just replaced `SMSG_UPDATE_OBJECT` with the compressed version in the vanilla server example and it seems to work. Removing that message entirely seems to just cause the client to get stuck at 100% on loading, which I'm assuming is what would happen if the compressed message was wrong.

**Comment by gtker** [comment]
> I hadn't tried that, but I just replaced SMSG_UPDATE_OBJECT with the compressed version in the vanilla server example

That's exactly what I meant. :)

> Removing that message entirely seems to just cause the client to get stuck at 100% on loading

Yeah, it's not possible to load in without the initial `SMSG_UPDATE_OBJECT`.

File: wow_message_parser/src/doc_printer/container.rs [hunk 5]
```diff
@@ -434,8 +434,26 @@ fn print_container_examples(s: &mut DocWriter, e: &Container, o: &Objects) {
...
+            // All fully compressed messages have a u32 at the start with the decompressed size.
+            let _decompressed_size: Vec<&u8> = bytes.clone().take(4).collect();
```
[the remark this answers is not in the document]
Comment (bigglesss) [reply wow_message_parser/src/doc_printer/container.rs:439]: Nope, looks like I completely forgot about that. Should it be printed at the start of the example? I can't see an easy way of just printing an additional custom member that isn't already part of the spec without duplicating code.

This sort of ties into your point about the examples being unclear. Arguably the examples shouldn't really include the uncompressed data at all. We don't really care about what the uncompressed raw data looks like (as it will be represented via a Rust struct anyway).

File: wow_message_parser/src/rust_printer/structs/print_common_impls/print_read.rs (hunk 3 above)
Comment (bigglesss) [reply wow_message_parser/src/rust_printer/structs/print_common_impls/print_read.rs:841]: It seems as though there's a `new_with_buf` method that takes a buffer, but it doesn't allow for *just* preallocation. I think it still makes sense to do this, as `decompressed_size` is *probably* using `compressBound` from the C ZLib lib, which is guaranteed to be larger than the actual compressed thing?

Comment (bigglesss) [reply wow_message_parser/src/rust_printer/structs/print_common_impls/print_read.rs:841]: Done

File: wow_message_parser/src/doc_printer/container.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bigglesss) [reply wow_message_parser/src/doc_printer/container.rs:439]: AFAIK the compressed messages all exist in uncompressed formats already, so perhaps the solution is to link to the uncompressed equivalent (COMPRESSED_UPDATE_OBJECT -> UPDATE_OBJECT, COMPRESSED_MOVES -> MONSTER_MOVE) and only show the compressed data on the actual page

[the remark this answers is not in the document]
Comment (gtker) [reply wow_message_parser/src/doc_printer/container.rs:439]: That's an interesting idea and very doable, but I don't want to lock ourselves in to a specific way of handling compressed messages in case a new message breaks this pattern. For now I think it'll be good enough if you don't uncompress and parse the example, but instead just parse the size u32 and add a comment and then parse the rest of the message.

Then manually in the `comment` tag write something like `Compressed version of [SMSG_UPDATE_OBJECT]. Has the same fields when uncompressed`.

Like so

```
0, 0, 0, 0, // Uncompresed size
0, 0, 0, [..],
0, 0, 0, 0, // Compressed data
```

File: wowm_language/src/docs/smsg_compressed_update_object.md (hunk 4 above)
Comment (gtker) [reply wowm_language/src/docs/smsg_compressed_update_object.md:34]: Examples are displayed compressed with u32 for size, and then manually linked to uncompressed version in the comments as described in the other conversation.

**Review by gtker — APPROVED** [review]
